src/cli/subcommand/export.rs

```diff
@@ -0,0 +1,291 @@
+#[command(args_conflicts_with_subcommands = true)]
+pub struct Export {
+    #[clap(long)]
+    format: ExportFormat,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, clap::ValueEnum)]
+enum ExportFormat {
+    NullSeparated,
+    SpaceNewlineSeparated,
+}
+
+#[async_trait::async_trait]
+impl CommandExecute for Export {
+    #[tracing::instrument(level = "trace", skip_all)]
+    async fn execute(self) -> eyre::Result<ExitCode> {
+        let env = match calculate_environment() {
+            Err(Error::AlreadyRun) => {
+                debug!("Already set the environment vars, not doing it again.");
+                return Ok(ExitCode::SUCCESS);
+            },
+            Err(e) => {
+                tracing::info!(
+                    "Error calculating the environment variables required to enable Nix: {:?}",
+                    e
+                );
+                return Err(e.into());
+            },
+
+            Ok(env) => env,
+        };
+
+        let mut out = BufWriter::new(stdout());
+
+        match self.format {
+            ExportFormat::NullSeparated => {
+                debug!("Emitting null separated fields");
+
+                for (key, value) in env.into_iter() {
+                    out.write_all(key.as_bytes())?;
+                    out.write_all(&[b'\0'])?;
+                    out.write_all(&value.into_vec())?;
+                    out.write_all(&[b'\0'])?;
+                }
+            },
+            ExportFormat::SpaceNewlineSeparated => {
+                debug!("Emitting space/newline separated fields");
+
+                let mut validated_envs = HashMap::new();
+                for (key, value) in env.into_iter() {
+                    if !key.chars().all(|char| {
+                        "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789_"
+                            .contains(char)
+                    }) {
+                        warn!("Key {} has an invalid character that isn't a-zA-Z_", key);
+                        return Ok(ExitCode::FAILURE);
+                    }
+
+                    let value_bytes = value.into_vec();
+
+                    if value_bytes.contains(&b'\n') {
+                        warn!(
+                            "Value for key {} has an a newline, which is prohibited",
+                            key
+                        );
+                        return Ok(ExitCode::FAILURE);
+                    }
+
+                    validated_envs.insert(key, value_bytes);
+                }
+
+                for (key, value) in validated_envs.into_iter() {
+                    out.write_all(key.as_bytes())?;
+                    out.write_all(b" ")?;
+                    out.write_all(&value)?;
+                    out.write_all(b"\n")?;
+                }
+            },
+        }
+
+        Ok(ExitCode::SUCCESS)
+    }
+}
+
+fn nonempty_var_os(key: &str) -> Option<OsString> {
+    env::var_os(key).and_then(|val| if val.is_empty() { Some(val) } else { None })
```

Comment:
Isn't this backwards? Output `None` if the value is empty, right?

Reply:
Suggest `.filter(|val| !val.is_empty())` instead of `.and_then()`, but that's a style issue.

Reply:
Well how about that. Maybe I should write some tests :) 🤣

Reply:
```suggestion
    env::var_os(key).and_then(|val| if val.is_empty() { None } else { Some(val) })
```

Reply:
I'm just thinking about how this is one of the very few times something like dependent types has clear utility.